axiom/nr_segment_datastore.c

```diff
@@ -107,6 +107,19 @@ bool nr_segment_datastore_end(nr_segment_t** segment_ptr,
 
   txn = segment->txn;
 
+  /*
+   * We don't want datastore segments to have any children, as
+   * this would scramble the exclusive time calculation.
+   *
+   * Therefore, we delete all children of the segment.
+   */
+  if (segment) {
+    for (size_t i = 0; i < nr_segment_children_size(&segment->children); i++) {
+      nr_segment_t* child = nr_segment_children_get(&segment->children, i);
```

Comment:
Is this loop going to work while you remove segments?  Seems like the index will be off as it is relative to the original size and it appears this code is removing elements?

Reply:
In this case no, we didn't actually remove the child from `children`, we just turned it NULL when we discarded the child segment so it is thereafter ignored. 
if we had instead used `nr_segment_children_remove` that would have changed the value of `nr_segment_children_size` each time through the loop and the incrementing `i` would have caused the index to be off.  However, `nr_segment_children_remove` doesn't discard the segment, it just goes through the expensive task of removing from the children vector after which we still have to discard.  So doing the above loop saves us the expense of removing the child and then discarding it because once we discard the segment, it is ignored.